### PR TITLE
Remove vertical seam

### PR DESCRIPTION
**Pull request by mfbx9da4**

### CODE REVIEW

File: SeamCarver.js [hunk 1]
```diff
@@ -163,7 +163,7 @@ class SeamCarver {
      */
     createEnergyMatrix() {
         // This has to be reverse order (bottom to top)
-        for (var y = this.height; y >= 0; y--) {
+        for (var y = this.height - 1; y >= 0; y--) {
```
Comment (mfbx9da4) [inline SeamCarver.js:181]: @ianribas you were actually creating another row!

File: SeamCarver.js [hunk 2]
```diff
@@ -203,13 +203,38 @@ class SeamCarver {
...
+     *
+     */
+    removeVerticalSeam(vseam) {
+        for (var row = 0; row < vseam.length; row ++) {
+            var deletedCol = vseam[row];
+
+            // Can ignore last column as we will delete it
+            for (var col = 0; col < this.width - 1; col ++) {
+                if (col >= deletedCol) {
+                    this.energy_matrix[col][row] = this.energy_matrix[col + 1][row];
```
Comment (mfbx9da4) [inline SeamCarver.js]: It seems to me this is a very inefficient way of traversing the matrix. If we wish to benefit from CPU caching we should either arrange the data structure into an array of rows or iterate each row for each col.

File: SeamCarver.js [hunk 3]
```diff
@@ -203,13 +203,38 @@ class SeamCarver {
...
+                }
+            }
+        }
+        this.energy_matrix.splice(this.width - 1, 1);
+        this.width--;
```
Comment (mfbx9da4) [inline SeamCarver.js:259]: Need to update `energy_matrix` next ...

File: SeamCarver.js (hunk 1 above)
Comment (ianribas) [inline SeamCarver.js:181]: Good catch. Off by one error :-(

File: SeamCarver.js (hunk 3 above)
Comment (ianribas) [inline SeamCarver.js:259]: One thing I was thinking is to add RGB info to the matrix and then just recalculate all again after removing the seam.

File: SeamCarver.js (hunk 2 above)
Comment (ianribas) [inline SeamCarver.js]: It does seem to me too that this is where we will need improving. For now, couldn't you use splice? In the off chance its implementation by the JS runtime is a little better...

File: readme.md [hunk 4]
```diff
@@ -16,3 +16,9 @@ Everyone
 ##Test
 
     npm run test
+
+##Optimizations
```
Comment (ianribas) [inline readme.md:25]: One more possible optimization I thought worth noting is to do logical deletes on the energy matrix. This way the cost of deletion goes way down, with the cost of finding the neighbor when recalculating going up some.

File: SeamCarver.js (hunk 2 above)
Comment (mfbx9da4) [inline SeamCarver.js]: Ah yes I see potentially, should time the difference...

File: SeamCarver.js [hunk 5]
```diff
@@ -86,22 +94,28 @@ class SeamCarver {
...
             return BORDER_ENERGY;
         }
-        var xant = this.context.getImageData(x - 1, y, 1, 1).data;
-        var xpost = this.context.getImageData(x + 1, y, 1, 1).data;
-        var yant = this.context.getImageData(x, y - 1, 1, 1).data;
-        var ypost = this.context.getImageData(x, y + 1, 1, 1).data;
 
-        return Math.sqrt(
+        var pos = this.pixelToIndex(x - 1, y);
+        var xant  = Array.prototype.slice.call(this.picture, pos, pos + 3);
```
Comment (mfbx9da4) [inline SeamCarver.js]: Could be quicker to iterate array rather than creating slices?

File: SeamCarver.js [hunk 6]
```diff
@@ -203,17 +220,105 @@ class SeamCarver {
...
+        this.energy_matrix.splice(this.width - 1, 1)
+        this.picture = this.imageData.data;
+        this.width--;
+
+        // now update energy matrix
+        // don't need to recalculate last row
+        for (var row = this.height - 2; row >= 0; row--) {
+            var deletedCol = vseam[row];
+
+            for (var i = -3; i < 4; i ++) {
```
Comment (ianribas) [inline SeamCarver.js]: I think that, after removing the seam, you actually have to recalculate the whole matrix! Or else  vminsum might be wrong. A change of the energy in one pixel propagate upwards to all three possible parents, that means that a change on the last row might (theoretically) impact the minimum sum of any pixel on the top row - not only those adjacent to the one that was removed ...

File: SeamCarver.js (hunk 5 above)
Comment (ianribas) [inline SeamCarver.js]: These `slice` calls really do make things slow. I profiled the demo. It's not iterating the array, just index adjustment. I'll make a pull request for this.

Comment (mfbx9da4) [inline SeamCarver.js]: Cool

**Pull request merged**